### DISCUSSION TITLE
How to set default value of React Select input from an async fetch response

### DISCUSSION THREAD

**Discussion by cliffordp** [react-hook-form/react-hook-form#2179] [Q&A] [answered]
The only way I've found to get _defaultValues_ working is if the function is outside the React component and it's a hard-coded value or something from the global _window_ object.

However, I need to set a value for a React Select field from an async fetch() response and it's not working even though it's in the same format as the defaultValues hard-coded one is.

I've been at this for a long while so any pointers would be really great. tyvm!

===

Because of having to be outside the component, I can't use React hooks like React Query so I just did a straight fetch()

![image](https://user-images.githubusercontent.com/1812179/87270820-edae8f00-c496-11ea-99ca-9869a719da56.png)

**Comment by bluebill1049** [comment]
have you tried by just using `setValue` and with `Controller`?

**Reply by cliffordp** [reply]
For sure. It fires the function (since it has a console.log) but it doesn't actually set the value.

**Reply by bluebill1049** [reply]
can you provide a simple CSB?

**Comment by cliffordp** [comment]
Correction, I see `defaultValue` as a Controller property but not `setValue` at https://react-hook-form.com/api#Controller

Here's my start: https://codesandbox.io/s/rhf-react-select-setvalue-98ns4?file=/src/index.js
Notice the `// Works:` type of comments

Find "getReactSelectDefault" in the code.

**Reply by bluebill1049** [reply]
i think you should use `reset({...getValues(), {xxxx} })`

**Reply by cliffordp** [reply]
Reset is part of Controller?

**Reply by bluebill1049** [reply]
yes.

**Comment by cliffordp** [comment]
Sorry, I'm not following... so in my sandbox... go from this:

```js
<Controller
              as={ReactSelect}
              options={options}
              name="ReactSelect"
              isClearable
              control={control}
              defaultValue={getReactSelectDefault()}
            />
```

to this?

```js
<Controller
              as={ReactSelect}
              options={options}
              name="ReactSelect"
              isClearable
              control={control}
              reset({...getReactSelectDefault(), {xxxx} })
            />
```

**Reply by bluebill1049** [reply]
no, sorry I misunderstood your question and confused you. just reset the whole form at the `useEffect` with `reset`

**Comment by cliffordp** [comment]
So upon initial form loading, reset entire form? What's the useEffect dependency? Have a sandbox link?

**Reply by bluebill1049** [reply]
after your request is resolved
```
reset({
  ...getValues(),
  xxxx // your select updated value
})
```

**Comment by cliffordp** [comment]
Sorry I'm needing extra help here :(

I updated [my CSB link](https://codesandbox.io/s/rhf-react-select-setvalue-98ns4?file=/src/index.js) as far along as I could get it but no dice.
I found the reset example at https://react-hook-form.com/api#reset but that's for resetting the entire form and I'm just trying to set the default value for a single ReactSelect field whenever the form loads, not to change someone's existing input if already touched.

Thanks much for the extra hand-holding here.

**Reply by bluebill1049** [reply]
yes, you can omit any of the formState changes.

**Comment by cliffordp** [comment]
I get that for _reset's_ args, but I'm not even getting that far along. Would you please look at the CSB and correct it for me?

**Reply by bluebill1049** [reply]
i can, can you fix the CSB, i will take a look later during lunch break
<img width="1282" alt="Screen Shot 2020-07-15 at 10 22 54 am" src="https://user-images.githubusercontent.com/10513364/87488995-28761a00-c685-11ea-93a2-e0e785db661f.png">

**Comment by cliffordp** [comment, marked as answer]
the `useEffect` snippet was doing all that - no error now that it's commented out

**Reply by bluebill1049** [reply]
this is what you mean: https://codesandbox.io/s/rhf-react-select-setvalue-v0u1z?file=/src/index.js?

**Comment by cliffordp** [comment]
omg tyvm!

**Reply by bluebill1049** [reply]
all good :)

**Comment by cliffordp** [comment]
Out of curiosity, why does using an async function (vs non-async) cause the following error?

> Warning: A component is changing a controlled input of type hidden to be uncontrolled. Input elements should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components

**Reply by bluebill1049** [reply]
because react need to know it's a controlled or uncontrolled during the render

**Reply by cubecleveland** [reply]
HI Guys,

Im struggline with this as it seems it should be a simple implementation.
Also trying to implement setting the option in the async after the fetch.......
Im not using a controller .
Here is my code:

const loadOptions = async (inputValue, callback) => {
       let i = await Site.search({name: inputValue}).then(e => e.data)        
       return i;
    }
<AsyncSelect
                    // styles={customStyles}
                      loadOptions={loadOptions}
                      ref={ref}
                      defaultOptions
                      isMulti
                      isDisabled={isAllSites}
                      getOptionLabel={e => e.name}
                      getOptionValue={e => e.id}
                      closeMenuOnSelect={false}

                      defaultValue={values ??? what should go in here}
                  />

**Reply by bluebill1049** [reply]
you should use `defaultValues` at the `useForm`.

**Comment by phoopwint15** [comment]
I want to reset my form with initial values fetched from the API. One field, vaccinename, depends on another field, vaccinetype. Right now, when I reset the form with API data, vaccinename doesn’t show the correct value because it’s tied to vaccinetype. I want vaccinename to show the API value initially, but only update when the user changes vaccinetype. How can I do this?

`  
  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      vaccinetype: "",
      vaccinename: "",
  });

  const selectedVaccineTypeName = useWatch({
    control: form.control,
    name: "vaccinetype",
  });
  const selectedVaccineType = vaccinetypes?.find(
    (v) => v.name === selectedVaccineTypeName,
  );
  const vaccineTypeId = Number(selectedVaccineType?.id) || undefined;
  const { data: vaccines } = useVaccinesByType(vaccineTypeId);

useEffect(() => {
    if (!vaccination) return;
    const timer = setTimeout(() => {
      form.reset({
        vaccinetype: vaccination.vaccine?.vaccine_type?.name ?? "",
        vaccinename: vaccination.vaccine?.id.toString(),
      });
    }, 300);
    return () => clearTimeout(timer);
  }, [vaccination, form]);

          <FormField
            control={form.control}
            name="vaccinetype"
            render={({ field }) => (
              <FormItem className="py-2.5 mr-2.5 w-full">
                <FormLabel required>Vaccine Type</FormLabel>
                <Select
                  onValueChange={(value) => {
                    field.onChange(value);
                    form.setValue("vaccinename", "");
                    form.trigger("vaccinename");
                  }}
                  value={field.value?.toString()}
                >
                  <FormControl className="w-full">
                    <SelectTrigger>
                      <SelectValue placeholder="Select vaccine type" />
                    </SelectTrigger>
                  </FormControl>
                  <SelectContent>
                    {vaccinetypes?.map((v) => (
                      <SelectItem key={v.uuid} value={v.name}>
                        {v.name}
                      </SelectItem>
                    ))}
                  </SelectContent>
                </Select>
                <FormMessage />
              </FormItem>
            )}
          />
          <div className="grid grid-cols-1 lg:grid-cols-2 gap-2.5">
            <FormField
              control={form.control}
              name="vaccinename"
              render={({ field }) => (
                <FormItem className="py-2.5 mr-2.5 w-full">
                  <FormLabel required>Vaccine Name</FormLabel>
                  <Select
                    onValueChange={field.onChange}
                    value={field.value?.toString()}
                  >
                    <FormControl className="w-full">
                      <SelectTrigger>
                        <SelectValue placeholder="Select vaccine name" />
                      </SelectTrigger>
                    </FormControl>
                    <SelectContent>
                      {vaccines?.map((v) => (
                        <SelectItem key={v.uuid} value={v.id.toString()}>
                          {v.name}
                        </SelectItem>
                      ))}
                    </SelectContent>
                  </Select>
                  <FormMessage />
                </FormItem>
              )}
            />
`
